text/1012-range-types.rst

```diff
@@ -0,0 +1,277 @@
+  values over the corresponding concrete ``anypoint`` sub-type
+
+
+Boundaries
+----------
+
+A ``range`` value should have an upper and lower bound and an indicator for
+each of them whether they are included in the values the ``range`` represents.
+
+It should be possible to omit one or both of the boundaries to represent an
+open-ended interval. The ``{}`` will be used to represent the omitted
+boundary. It should be an error to omit the boundary, but indicate that it is
+inclusive.
+
+
+Constructors
+------------
+
+We will use a function constructor for creating new ``range`` values::
+
+  function range(
+      lower: optional anypoint = {},
+      upper: optional anypoint = {},
+      named only inclow: bool = true,
+      named only inchigh: bool = false
+  ) -> range<anypoint>
+
+The default values are chosen so as to minimize the amount of keystrokes
+required to define ranges for common use-cases such as for iterators.
+
+The ``inclow`` and ``inchigh`` arguments are ``named only`` because when their
+default values need to be overridden it is entirely plausible that only one of
+them needs to be specified and thus using positional arguments would be
+awkward.
+
+A ``multirange`` could be constructed using a function that takes a set of
+basic ``range`` values::
+
+  function multirange(
+      ranges: set of range<anypoint>
+  ) -> multirange<anypoint>
+
+The constructor must normalize the ranges to be non-overlapping and order them
+based on boundaries. This will ensure that when a ``multirange`` is unpacked
+into its component ranges, the order of the ``range`` values would be based on
+their boundaries rather than arbitrary.
+
+
+Accessing range components
+--------------------------
+
+Several functions will be used to access ``range`` components::
+
+  function get_range_upper(range<anypoint>) -> optional anypoint
+  function get_range_lower(range<anypoint>) -> optional anypoint
+  function is_inclusive_lower(range<anypoint>) -> bool
+  function is_inclusive_upper(range<anypoint>) -> bool
+
+
+Accessing multirange components
+-------------------------------
+
+There will be a function for extracting sub-ranges from a multirange::
+
+  function get_subranges(multirange<anypoint>) -> set of range<anypoint>
+
+By default this function will return the ranges ordered by boundaries
+fromlowest to highest. This is the order of definition and Postgres by default
+returns the sub-intervals in that order.
+
+
+Helper functions
+----------------
+
+In addition to the functions mentioned above there must be a way to generate
+an ordered set of values from a ``range`` or ``multirange``::
+
+  function range_unpack(
+      val: range<anydiscrete>,
+      step: anydiscrete = DEFAULT_STEP
+  ) -> set of anydiscrete
+
+  function range_unpack(
+      val: multirange<anydiscrete>,
+      step: anydiscrete = DEFAULT_STEP
+  ) -> set of anydiscrete
+
+
+  function range_unpack(
+      val: range<anycontiguous>,
+      step: anycontiguous
```

Comment:
We would have to invent another type: `increment<anycontiguous>` (and `increment<anydiscrete>`), which would internally resolve into the appropriate increment type of a given point type, i.e. `duration` for `datetime`.

Reply:
I don't think that inventing another type makes any sense here because for `local_date`, `local_time`, and `datetime` the step is `duration`, but it's also a valid step for `duration` itself. Trying to link them by marking them with abstract types doesn't seem to work because abstract types don't let us see the valid pairs.

Date and time scalars seem to be the only current exception, but if we ever have another one, generalizing this mechanism would be near impossible. On the other hand, encoding this feature as a particular overload for `range_unpack` is clear and future-proof.

Reply:
`increment<type>` is not "another type", it's an alias for the return type of operator `-(type, type)`, i.e. it needs no actual formal implementation aside from a bit of schema magic.

Reply:
Wait, are you suggesting that `increment<type>` is a type expression that isn't independently defined, but instead literally looks up the operator `-(type, type)` and returns the declared return type of that operator?

And the main reason for this thing even existing is to be used in formal generic function definitions like `range_unpack(val: range<anypoint>, step: increment<anypoint>)`? What's the benefit? We would still implement the specific overloaded versions for each specific range type, I think.

If it is independently defined, then it would place the burden of reconciling it with the operator, so maybe not a great option.

Reply:
> but instead literally looks up the operator `-(type, type)` and returns the declared return type of that operator?

This.

> We would still implement the specific overloaded versions for each specific range type, I think.

If we implement the "magic" type then we wouldn't necessarily need to implement the overloads, because the compiler can just assume that all descendants of `anypoint` implement the necessary generator.

Reply:
Added `increment<type>` to the RFC.